kaldb/src/main/java/com/slack/kaldb/preprocessor/PreprocessorService.java

```diff
@@ -0,0 +1,285 @@
+  private final List<String> upstreamTopics;
+  private final String downstreamTopic;
+  private final String dataTransformer;
+  private final MeterRegistry meterRegistry;
+
+  private KafkaStreams kafkaStreams;
+  private KafkaStreamsMetrics kafkaStreamsMetrics;
+
+  private final Counter configReloadCounter;
+  public static final String CONFIG_RELOAD_COUNTER = "preprocessor_config_reload_counter";
+
+  public PreprocessorService(
+      ServiceMetadataStore serviceMetadataStore,
+      KaldbConfigs.PreprocessorConfig preprocessorConfig,
+      MeterRegistry meterRegistry) {
+    this.serviceMetadataStore = serviceMetadataStore;
+    this.meterRegistry = meterRegistry;
+    this.configReloadCounter = meterRegistry.counter(CONFIG_RELOAD_COUNTER);
+
+    this.kafkaProperties = makeKafkaStreamsProps(preprocessorConfig.getKafkaStreamConfig());
+    this.downstreamTopic = preprocessorConfig.getDownstreamTopic();
+    this.upstreamTopics = Collections.unmodifiableList(preprocessorConfig.getUpstreamTopicsList());
+    this.dataTransformer = preprocessorConfig.getDataTransformer();
+    this.rateLimiter =
+        new PreprocessorRateLimiter(
+            meterRegistry, preprocessorConfig.getPreprocessorInstanceCount());
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting preprocessor service");
+    load();
+    serviceMetadataStore.addListener(this::load);
+    LOG.info("Preprocessor service started");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Stopping preprocessor service");
+    if (kafkaStreams != null) {
+      kafkaStreams.close(DEFAULT_START_STOP_DURATION);
+    }
+    if (kafkaStreamsMetrics != null) {
+      kafkaStreamsMetrics.close();
+    }
+    serviceMetadataStore.removeListener(this::load);
+    LOG.info("Preprocessor service closed");
+  }
+
+  /**
+   * Configures and starts a KafkaStream processor, based off of the cached ServiceMetadataStore.
+   * This method is reentrant, and will restart any existing KafkaStream processors. Access to this
+   * must be synchronized if using this method as part of a listener.
+   */
+  public synchronized void load() {
+    configReloadCounter.increment();
+    LOG.info("Loading new Kafka stream processor config");
+    if (kafkaStreams != null) {
+      LOG.info("Closing existing Kafka stream processor");
+      kafkaStreams.close();
+      kafkaStreams.cleanUp();
+    }
+    if (kafkaStreamsMetrics != null) {
+      kafkaStreamsMetrics.close();
+    }
+
+    // only attempt to register stream processing on valid service configurations
+    List<ServiceMetadata> serviceMetadataToProcess =
+        filterValidServiceMetadata(serviceMetadataStore.listSync());
+
+    if (serviceMetadataToProcess.size() > 0) {
+      Topology topology =
+          buildTopology(
+              serviceMetadataToProcess,
+              rateLimiter,
+              upstreamTopics,
+              downstreamTopic,
```

Comment:
I think multi-tenancy would be hard to support if both topics should be on the same kafka cluster.  https://stackoverflow.com/questions/53034610/can-kafka-streams-output-topic-be-on-a-separate-cluster

Reply:
As discussed in https://github.com/slackhq/kaldb/pull/245#discussion_r847828855 we'll be addressing this in the followup PR.